### PR TITLE
Fix/1208238364944982 Generate DID Hash Offline

### PR DESCRIPTION
**Pull request by jpgundrum**
When a user wants to generate a DID Document hash this new updates allows them to do so offline. Using the SDK.generateDidDocument() function a did hash value is returned back which can be used to manually set the value.

### CODE REVIEW

File: packages/sdk/src/modules/main/index.ts [hunk 1]
```diff
@@ -14,19 +14,33 @@ import { RBAC } from "../rbac";
...
   private readonly _options: Options;
-  protected override _api: ApiPromise;
+  protected override _api: ApiPromise | undefined;
   private _metadata: SDKMetadata;
+
   public did: Did;
   public rbac: RBAC;
+  static offlineMode: boolean;
 
   constructor(options: Options) {
```
Comment (jpgundrum) [inline packages/sdk/src/modules/main/index.ts:25]: @irediaes Please see how the constructor changes.

File: packages/sdk/src/modules/main/index.ts [hunk 2]
```diff
@@ -14,19 +14,33 @@ import { RBAC } from "../rbac";
  */
 export class Main extends Base {
   private readonly _options: Options;
-  protected override _api: ApiPromise;
+  protected override _api: ApiPromise | undefined;
```
Comment (irediaes) [inline packages/sdk/src/modules/main/index.ts]: revert back to previous.
remove "undefined"

File: packages/sdk/src/modules/main/index.ts [hunk 3]
```diff
@@ -14,19 +14,33 @@ import { RBAC } from "../rbac";
...
   private _metadata: SDKMetadata;
+
   public did: Did;
   public rbac: RBAC;
+  static offlineMode: boolean;
```
Comment (irediaes) [inline packages/sdk/src/modules/main/index.ts:23]: remove please.
Check my suggested implementation below

File: packages/sdk/src/modules/main/index.ts [hunk 4]
```diff
@@ -14,19 +14,33 @@ import { RBAC } from "../rbac";
...
+      super();
+      this._options = options;
+      this._api = this._createApi(options);
+      this._metadata = {};
 
-    this.did = new Did(this._api, this._metadata);
-    this.rbac = new RBAC(this._api, this._metadata);
+      this.did = new Did(this._api, this._metadata);
+      this.rbac = new RBAC(this._api, this._metadata);
+    }
```
Comment (irediaes) [inline packages/sdk/src/modules/main/index.ts]: Too many messy hacks.

Please try this instead:

```
constructor(options: Options) {
   super();
    this._options = options;
    this._api = this._createApi(options);
    this._metadata = {};
    
    this.did = new Did(this._api, this._metadata);
    this.rbac = new RBAC(this._api, this._metadata);
}
```
Then createInstance():

```
 public static async createInstance(options: Options): Promise<Main> {
    await cryptoWaitReady();
    const sdk = new Main(options);
    await sdk.connect();
    return sdk;
  }
```

Change `createOfflineInstance` to `generateDidDocument` in the Main class:
Other irrelevant did methods won't be exposed if implemented this way.

```

public static async generateDidDocument(options: GenerateDidOptions): Promise<GenerateDidResult> {
      const did = new Did();
      return did.generateDidDocument(options);
}
```

so usage will be:

```
import {Sdk} from `@peaq-network/sdk

const didVal = await Sdk.generateDidDocument({name:"DID"});

```

**Review by irediaes — CHANGES_REQUESTED** [review]
Too many messy hacks.

Please try this instead:

```
constructor(options: Options) {
   super();
    this._options = options;
    this._api = this._createApi(options);
    this._metadata = {};
    
    this.did = new Did(this._api, this._metadata);
    this.rbac = new RBAC(this._api, this._metadata);
}
```
Then createInstance():

```
 public static async createInstance(options: Options): Promise<Main> {
    await cryptoWaitReady();
    const sdk = new Main(options);
    await sdk.connect();
    return sdk;
  }
```

Change `createOfflineInstance` to `generateDidDocument` in the Main class:
Other irrelevant did methods won't be exposed if implemented this way.

```

public static async generateDidDocument(options: GenerateDidOptions): Promise<GenerateDidResult> {
      const did = new Did();
      return did.generateDidDocument(options);
}
```

so usage will be:

```
import {Sdk} from `@peaq-network/sdk

const didVal = await Sdk.generateDidDocument({name:"DID"});

```

Let me know what you think.

File: packages/sdk/src/modules/main/index.ts (hunk 4 above)
Comment (jpgundrum) [reply packages/sdk/src/modules/main/index.ts]: This works. Thanks.

**Comment by jpgundrum** [comment]
Closing this pull request to rename the branch to a more appropriate name.

**Pull request closed without merging**